crates/p2p/src/swarm/mod.rs

```diff
@@ -432,6 +455,59 @@ impl P2P {
     async fn handle_swarm_event(&mut self, event: SwarmEvent<BehaviourEvent>) -> P2PResult<()> {
         match event {
             SwarmEvent::Behaviour(event) => self.handle_behaviour_event(event).await,
+            SwarmEvent::ConnectionEstablished { .. }
+            | SwarmEvent::OutgoingConnectionError { .. } => {
+                self.handle_connection_event(event).await
+            }
+            _ => Ok(()),
+        }
+    }
+
+    /// Handles connection-related events (both successful and failed connections).
+    async fn handle_connection_event(
+        &mut self,
+        event: SwarmEvent<BehaviourEvent>,
+    ) -> P2PResult<()> {
+        match event {
+            SwarmEvent::ConnectionEstablished {
+                peer_id,
+                connection_id,
+                ..
+            } => {
+                if let Some(app_public_key) = self
+                    .dial_manager
+                    .get_app_public_key_by_connection_id(&connection_id)
+                    .await
+                {
+                    self.dial_manager.remove_queue(&app_public_key).await;
+                    self.dial_manager.remove_connid(&connection_id).await;
+                    info!(peer_id = %peer_id, "connected to peer");
+                }
+                Ok(())
+            }
+            SwarmEvent::OutgoingConnectionError {
+                connection_id,
+                error,
+                ..
+            } => {
+                if let Some(app_public_key) = self
+                    .dial_manager
+                    .get_app_public_key_by_connection_id(&connection_id)
+                    .await
+                {
+                    warn!(app_public_key = ?app_public_key, error = %error, "connection failed");
+                    self.dial_manager.remove_connid(&connection_id).await;
+                    if let Some(next_addr) = self.dial_manager.pop_next_addr(&app_public_key).await
+                    {
+                        info!(next_addr = %next_addr, app_public_key = ?app_public_key, "retrying with next address");
+                        self.dial_and_map(next_addr, app_public_key).await;
+                    } else {
+                        warn!(app_public_key = ?app_public_key, "no more addresses to try, removing queue");
+                        self.dial_manager.remove_queue(&app_public_key).await;
+                    }
```

Comment:
Deadlock Risk: This creates even more complex lock acquisition patterns:

  - `get_app_public_key_by_connection_id()`
  - `remove_connid()`
  - `pop_next_addr()` (conditional)
  - `remove_queue()` (conditional in else branch)